datafusion/physical-plan/src/sorts/partial_sort.rs

```diff
@@ -0,0 +1,1005 @@
+    }
+
+    fn execute(
+        &self,
+        partition: usize,
+        context: Arc<TaskContext>,
+    ) -> Result<SendableRecordBatchStream> {
+        trace!("Start PartialSortExec::execute for partition {} of context session_id {} and task_id {:?}", partition, context.session_id(), context.task_id());
+
+        let input = self.input.execute(partition, context.clone())?;
+
+        trace!(
+            "End PartialSortExec's input.execute for partition: {}",
+            partition
+        );
+
+        // Make sure common prefix length is larger than 0
+        // Otherwise, we should use SortExec.
+        assert!(self.common_prefix_length > 0);
+
+        Ok(Box::pin(PartialSortStream {
+            input,
+            expr: self.expr.clone(),
+            common_prefix_length: self.common_prefix_length,
+            input_batch: RecordBatch::new_empty(self.schema().clone()),
+            fetch: self.fetch,
+            is_closed: false,
+            baseline_metrics: BaselineMetrics::new(&self.metrics_set, partition),
+        }))
+    }
+
+    fn metrics(&self) -> Option<MetricsSet> {
+        Some(self.metrics_set.clone_inner())
+    }
+
+    fn statistics(&self) -> Result<Statistics> {
+        self.input.statistics()
+    }
+}
+
+struct PartialSortStream {
+    /// The input plan
+    input: SendableRecordBatchStream,
+    /// Sort expressions
+    expr: Vec<PhysicalSortExpr>,
+    /// Length of
+    common_prefix_length: usize,
+    /// Used as a buffer for part of the input not ready for sort
+    input_batch: RecordBatch,
+    /// Fetch top N results
+    fetch: Option<usize>,
+    /// Whether the stream has finished returning all of its data or not
+    is_closed: bool,
+    /// Execution metrics
+    baseline_metrics: BaselineMetrics,
+}
+
+impl Stream for PartialSortStream {
+    type Item = Result<RecordBatch>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        let poll = self.poll_next_inner(cx);
+        self.baseline_metrics.record_poll(poll)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        // we can't predict the size of incoming batches so re-use the size hint from the input
+        self.input.size_hint()
+    }
+}
+
+impl RecordBatchStream for PartialSortStream {
+    fn schema(&self) -> SchemaRef {
+        self.input.schema()
+    }
+}
+
+impl PartialSortStream {
+    fn poll_next_inner(
+        self: &mut Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<RecordBatch>>> {
+        if self.is_closed {
+            assert_eq!(0, self.input_batch.num_rows());
+            return Poll::Ready(None);
+        }
+        let result = match ready!(self.input.poll_next_unpin(cx)) {
+            Some(Ok(batch)) => {
+                self.input_batch = compute::concat_batches(
+                    &self.schema(),
+                    &[self.input_batch.clone(), batch],
```

Comment:
```suggestion
                    [&self.input_batch, &batch],
```
concat_batches util accepts references also.